### ISSUE TITLE
G>O recurring series continually creates duplicates before sync date window

### ISSUE THREAD

**Issue by hikaritenchi** [phw198/OutlookGoogleCalendarSync#1910]
> Checklist:-
> * [x] [Upgrade](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Upgrading) to the latest release (alpha if available) ![latest release](https://img.shields.io/github/release/phw198/OutlookGoogleCalendarSync/all.svg)
> * [x] Reproduce the issue
> * [x] Complete the information below
> * [x] Provide your `OGcalsync.log` file (see [how-to](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Reporting-problems#locating-your-ogcalsynclog-file))

**OGCS Version**: v2.10.6.0
**Edition**: Installed
**Sync Direction**: G->O

**Bug Description**  
When syncing, there is a calendar entry that continually gets recreated, and I don't really know why. 
It feels like the newly-created event doesn't get detected by OGCS, so it recreates it.
In the log file, you can see:
Sync 1: 80 Outlook calendar entries found, 81 Google calendar entries found
OGCS correctly creates the missing event
Sync 2: 80 Outlook calendar entries found, 81 Google calendar entries found
OGCS doesn't detect the newly-created event and recreates it, etc.

I had to turn off automatic sync because I had 30 of these events to delete from Outlook, haha.

**Steps to Reproduce the Issue**  
Not sure if it's possible to reproduce, just feels like this might be an edge case. If needed for debugging, I can try and create a public calendar with just one event, but this is an event that I'm invited to and not my own.

**Any other information**  
Any additional information that may help troubleshoot the issue.  
:paperclip: Attach your `OGcalsync.log` file.
I'll fill out the form to upload to the Google Drive.
The relevant logs are from `2024-07-18 08:51:25,580` until `2024-07-18 08:53:58,986`

**Comment by phw198** [comment]
The item that is being created is a recurring series starting on 12-Jun-2024 and _should_ end on 26-Jun-2024,_ which is outside the sync window of 11-Jul-2024 to 17-Sep-2024.

Can you confirm what the series looks like in Google? Does it show as continuing past 26-Jun? In fact, it looks like the occurrence on the 26th is an exception - was it moved to a date after 11-Jul, or deleted or something?

Which is the exact calendar item on which day, that shows as duplicating in Outlook?

**Comment by hikaritenchi** [comment]
The calendar invite in Google says
`Every 2 weeks on Wednesday, until Nov 2, 2024`
The event on the 26-June seems like it's a normal event, but there's one event that was originally scheduled for 10-July and got moved to 12-July.

The one being duplicated is the one on 12-June.

Thanks for responding so quickly!

**Comment by phw198** [comment]
Something doesn't seem right, maybe on the Google side. OGCS is seeing the event series with a recurrence definition until 26-Jun, but it's also retrieving an event exception for that date, which indicates it's been changed somehow (eg moved or deleted). I think it maybe was moved to a date that is within the sync window, which is causing it to sync.

#1884 also had a weird situation where the Google recurring series wasn't correct...could you maybe [follow the steps](https://github.com/phw198/OutlookGoogleCalendarSync/issues/1884#issuecomment-2178191045) to work out the Event ID of the item that you can see on the 12-Jun? OGCS sees it as `k20882b932losgerulh595iatp`

**Comment by hikaritenchi** [comment]
Thanks for the help!

Event{eid=k20882b932losgerulh595iatp_20240612T140000Z, organizer=electriclawncompany@gmail.com, participant=hikaritenchi@gmail.com, actor=hikaritenchi@gmail.com, summary=[REDACTED] | Electric Lawn Company , status=CONFIRMED, seq=3, startTime=2024-06-12T14:00:00Z, endTime=2024-06-12T14:50:00Z, firstStart=20240612T100000, rdata=RRULE:FREQ=WEEKLY;UNTIL=20241102T035959Z;INTERVAL=2;BYDAY=WE;WKST=SU}

Seems like the event ID matches, but there's a `Z` at the end of the Google one, but it seems like all the events have that.

**Comment by phw198** [comment]
It seems like there have been lots of changes to this recurring series (maybe the series start date got changed and then change made on 26-Jun that affects "this event and following"?) and things have got a bit messed up:

* Outlook has a recurring series "master", which is related to the Google event on 26-Jun
* In Google the recurring series "master" is on 12-Jun and the series apparently ends on 26-Jun
* However, according to your info posted above, the 12-Jun series doesn't end until 2-Nov in Google

I think the simplest next step is for you to delete all series and occurrences of the item in Outlook and let OGCS recreate it on the next sync - see if it behaves better after that, or the issue remains.

**Comment by hikaritenchi** [comment]
Thanks for the explanations, that makes sense. I did delete all the events in the series, then tried to sync, but had an error. I aborted that sync, then tried syncing again, which seemed to work. Upon the 3rd sync attempt, the duplicate event was still being created.

I've submitted another log, let me know if that helps with debugging at all.

**Comment by hikaritenchi** [comment]
Hi, just wanted to let you know that my lawn care provider updated the calendar invite, so now a duplicate event doesn't get created anymore. If this ends up happening again, I'll probably re-open this to keep the context.

Let me know if you'd like the log for the calendar update that removed this behavior.

**Comment by phw198** [comment]
Glad it's resolved for you. I've managed to replicate the issue and I'm going to put this down to a bug on Google's side! Essentially, when a series is changed for "this and future events", the future events form a new series that starts from the first changed occurrence. However,  Google still returns the original recurring series - even if it finished before the sync window.

:package: This should now be resolved/worked around with [hotfix v2.10.6.1.zip](https://github.com/user-attachments/files/16455158/v2.10.6.1.zip) ([see instructions for applying](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Reporting-Problems#applying-a-hotfix)).
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [CRCs](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Known-Issues#cyclic-redundancy-checks---aka-crcs): Zip = `929EA34A` Exe = `8FF8B383`

If anyone gives this a try, please let me know how it goes.

**Comment by hikaritenchi** [comment]
I wanted to say that I did run into a similar issue with another recurring event, and the hotfix does solve the problem in that duplicate events are no longer being created ad nauseam in Outlook. Looking forward to a more permanent fix!

**Comment by jacktose** [comment]
I'm having what looks like the same problem. Currently on v 2.11.2.1, but I think it dates back at least two upgrades.

It's definitely a long-running series in Google, in which I have frequently changed individual and "all following" events. The original is on 2021-10-21 or -28. Every time OGCS runs, it creates an Outlook event on 2021-10-21. Today, I deleted over 5,000 events. (My Outlook calendar is much snappier!)

Here's the (anonymized) GCal event info:
```js
Event{eid=lk92gct42jrv31fn58t5nniejn_20211028T233000Z, organizer=family01160251561657494786@group.calendar.google.com, participant=j***********@gmail.com, actor=j***********@gmail.com, summary=Bonnie, status=CONFIRMED, seq=4, startTime=2021-10-28T23:30:00Z, endTime=2021-10-29T00:30:00Z, firstStart=20211021T153000, rdata=RRULE:FREQ=WEEKLY;BYDAY=TH}
```

I've uploaded my log to the Google Drive. Clear examples are the two syncs between 2025-01-28 16:47:36 and 2025-01-28 16:54:23.

**Comment by jacktose** [comment]
I'm having what looks like the same problem: two events are recreated on every G->O sync.
I just deleted ~3,000 of each. (My Outlook calendar is much snappier!)

They're both long-running series in Google, in which I have frequently changed individual and "all following" events. The original of one is on 2021-10-21 or -28. Every time OGCS runs, it creates an Outlook event on 2021-10-21. The original of the other is on 2023-06-29, and every sync creates an Outlook event then.
In the GUI, this looks like:
> Creating 2 Outlook calendar entries
> 2021-10-21 23:30  "B*****?"
> 2023-06-30 03:00  "G*****************************y"

(Note that those times are wrong by my GMT offset. So were the actual events, while I was on v2.11.2.0.)

Here's the (anonymized) GCal event info:
```js
Event{eid=lk92gct42jrv31fn58t5nniejn_20211028T233000Z, organizer=family01160251561657494786@group.calendar.google.com, participant=j***********@gmail.com, actor=j***********@gmail.com, summary=B****e, status=CONFIRMED, seq=4, startTime=2021-10-28T23:30:00Z, endTime=2021-10-29T00:30:00Z, firstStart=20211021T153000, rdata=RRULE:FREQ=WEEKLY;BYDAY=TH}
Event{eid=o2mrdju4j1lefujtqc4p9qrhs8_20230630T030000Z, organizer=65i1m1csgo5h6gfl4i22bnjqp0@group.calendar.google.com, participant=j***********@gmail.com, actor=j***********@gmail.com, summary=G*****************************y, status=CONFIRMED, seq=4, startTime=2023-06-30T02:00:00Z, endTime=2023-06-30T04:00:00Z, firstStart=20230629T200000, rdata=RRULE:FREQ=WEEKLY;WKST=MO}
```

I've uploaded my log to the Google Drive. Clear examples are the two syncs between 2025-01-28 16:47:36 and 2025-01-28 16:54:23. [I put anonymized excerpts of those in a gist, too.](https://gist.github.com/jacktose/0283b502870918c760791c5abf4b3608) GCal also has a duplicate of the B****e event in the Family calendar, which OGCS syncs the same way. So I had another ~3,000 of those at the same time in Outlook.

Currently on v2.11.2.1, but I think the behavior dates back at least two upgrades (3k events @ 1 sync/hr ≈ 4mo.). @phw198, should I assume that hotfix v2.10.6.1 is included in v2.11.x? Or what version/hotfix should I use to avoid this issue for now? OGCS is effectively unusable for me with this bug.

**Comment by phw198** [comment]
You're right, v2.10.6.1 was never brought into the beta v2.11.0 because this fix is just part of a huge problem (splitting series in Google, but not Outlook) that I need to properly think about how to solve. So at the moment, I'd prefer to apply this as a rolling patch for each release until I can devise a plan and complete solution.

:package: With that in mind, here is [hotfix v2.11.2.5.zip](https://github.com/user-attachments/files/18606403/v2.11.2.5.zip) with the latest releases included (see [instructions for applying](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Reporting-Problems#applying-a-hotfix)).
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [CRCs](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Known-Issues#cyclic-redundancy-checks---aka-crcs): Zip = `E0D32719` Exe = `C10BAE05`

**Comment by jacktose** [comment]
Thanks! Installed v2.11.2.5 and test syncs look good (though times in OGCS UI output still seem offset). I've re-enabled scheduled syncs on all profiles and I'll keep an eye on it.
I suppose I'll just decline updates until I see the real fix in release notes.

(I wish Outlook would handle changes to series more like Google does (at least for the UX), but what can we do?)

**Comment by jacktose** [comment]
Is this hotfix incorporated into later versions?
I'm still running v2.11.2.5 because I'm scared of this issue reappearing. I'd like to update for some of your recent fixes (especially #2135), but I don't see any mention of this issue number or description on the releases page.

**Comment by phw198** [comment]
@jacktose Sorry it's taking so long to get around to a proper fix - happy for you to bug me for a patched version when you need it. 

:package: Please upgrade to v2.11.6 first, then apply [hotfix v2.11.6.3.zip](https://github.com/user-attachments/files/22984340/v2.11.6.3.zip) (see [instructions for applying](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Reporting-Problems#applying-a-hotfix)).
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [CRCs](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Known-Issues#cyclic-redundancy-checks---aka-crcs): Zip = `FBD4D126` Exe = `EE60E27F`

**Comment by jacktose** [comment]
Thank you! Successfully installed, and sync running successfully* with no duplication or TZ problems.

\* It does seem to be failing to sync a recurring event (with or without exceptions) G->O, but that's not this issue. If I can confirm, should I report in #2017 or #2038 maybe?

**Comment by phw198** [comment]
Glad you've got the patch installed OK 👍🏻 

I think just raise a new issue for the missing recurring series (both those you linked to are for v3 of OGCS and you're still on v2). Do double check you are not excluding and filtering out the series somehow though.

**Comment by phw198** [comment]
:package: Reposting from #2212 hotfix [v2.11.6.4.zip](https://github.com/user-attachments/files/23432441/v2.11.6.4.zip) (see [instructions for applying](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Reporting-Problems#applying-a-hotfix)).
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [CRCs](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Known-Issues#cyclic-redundancy-checks---aka-crcs): Zip = `49E736C7` Exe = `3153B889`

**Comment by phw198** [comment]
:package: Please upgrade to v2.11.7 first, then apply [hotfix v2.11.7.1.zip](https://github.com/user-attachments/files/24402194/v2.11.7.1.zip) (see [instructions for applying](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Reporting-Problems#applying-a-hotfix)).
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [CRCs](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Known-Issues#cyclic-redundancy-checks---aka-crcs): Zip = `AFAD13A5` Exe = `697F1081`

**Comment by andreas-kellermeier** [comment]
> 📦 Please upgrade to v2.11.7 first, then apply [hotfix v2.11.7.1.zip](https://github.com/user-attachments/files/24402194/v2.11.7.1.zip) (see [instructions for applying](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Reporting-Problems#applying-a-hotfix)).       [CRCs](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Known-Issues#cyclic-redundancy-checks---aka-crcs): Zip = `AFAD13A5` Exe = `697F1081`

The hotfix works. Thank you.

**Comment by phw198** [comment]
Decided to include this fix/workaround in the next beta release and assess the complete solution required under #2253.

**Comment by andreas-kellermeier** [comment]
👍❤️🤘

Viele Grüße,
Andreas

**Comment by andreas-kellermeier** [comment]
Hi Paul,

a question for you.

Have the sync errors already been fixed in version 3.0.0.35?

Best regards, Andreas

**Comment by phw198** [comment]
@andreas-kellermeier No, only in v2.12.0 release so far. It's going into the v3.0.1 alpha though which, if you're lucky, might be out this weekend... 🙂

**Comment by andreas-kellermeier** [comment]
👍

Viele Grüße,
Andreas

**Comment by andreas-kellermeier** [comment]
Is V 3.0.1 out now?

Viele Grüße,
Andreas

**Comment by phw198** [comment]
@andreas-kellermeier [v3 is out](https://github.com/phw198/OutlookGoogleCalendarSync/releases/tag/v3.0.1-alpha) in the last few hours 🙂 🎉

**Comment by andreas-kellermeier** [comment]
Looks very nice Paul

Viele Grüße,
Andreas
